ros/src/tl_detector/tl_detector.py

```diff
@@ -131,21 +149,44 @@ def process_traffic_lights(self):
             int: ID of traffic light color (specified in styx_msgs/TrafficLight)
 
         """
-        light = None
-
-        # List of positions that correspond to the line to stop in front of for a given intersection
+        closest_light = None
+        closest_light_wp = None
+        closest_light_stop_wp = None
+        dist_to_light = 10000   #initialize to high value
+	first = True
         stop_line_positions = self.config['stop_line_positions']
-        if(self.pose):
-            car_position = self.get_closest_waypoint(self.pose.pose)
-
-        #TODO find the closest visible traffic light (if one exists)
-
-        if light:
-            state = self.get_light_state(light)
-            return light_wp, state
-        self.waypoints = None
+        if(self.pose):		
+            	car_position = self.get_closest_waypoint(self.pose.pose)
+        else:
+        	return -1, TrafficLight.UNKNOWN
+
+        for i, light in enumerate(self.lights):
+		light_stop_pose = Pose()
+            	light_stop_pose.position.x = stop_line_positions[i][0]
+            	light_stop_pose.position.y = stop_line_positions[i][1]
+            	light_stop_wp = self.get_closest_waypoint(light_stop_pose)    #get the wp closest to each light_position
+
+            	if light_stop_wp >= car_position :    #if waypoint is  close to the traffic light and ahead of the car
+            		if first:    #check for the first light 
+				closest_light_stop_wp = light_stop_wp
+                    		closest_light = light
+				first = False
+                	elif light_stop_wp < closest_light_stop_wp:
+                    		closest_light_stop_wp = light_stop_wp  
+                    		closest_light = light
+
+        if ((car_position is not None) and (closest_light_stop_wp is not None)):
+        	dist_to_light = abs(car_position - closest_light_stop_wp)
+
+        if closest_light and dist_to_light < 50: #we check the status of the traffic light if it's within 50 waypoints	
```

Comment:
if the condition above is true `closest_light` is always not None so we can simplify the code:

```
if closest_light_stop_wp is not None:
    dist_to_light = abs(car_position - closest_light_stop_wp)
    if dist_to_light < 50: #we check the status of the traffic light if it's within 50 waypoints    
        state = self.get_light_state(closest_light)
        return closest_light_stop_wp, state
return -1, TrafficLight.UNKNOWN
```

Reply:
IMO, the value 50 should be in a variable / const